cachi2/core/package_managers/yarn/project.py

```diff
@@ -126,3 +142,46 @@ def yarn_cache(self) -> RootedPath:
     def from_source_dir(cls, source_dir: RootedPath) -> "Project":
         """Create a Project from a sources directory path."""
         return NotImplemented
+
+
+def get_semver_from_yarn_path(yarn_path: str) -> Optional[semver.version.Version]:
+    """Parse yarnPath from yarnrc and return a semver Version if possible else None."""
+    # https://github.com/yarnpkg/berry/blob/2dc59443e541098bc0104d97b5fc452781c64baf/packages/plugin-essentials/sources/commands/set/version.ts#L208
+    yarn_spec_pattern = re.compile(r"^yarn-(.+)\.cjs$")
+    match = yarn_spec_pattern.match(Path(yarn_path).name)
+    if not match:
+        return None
+
+    yarn_version = match.group(1)
+    try:
+        return semver.version.Version.parse(yarn_version)
+    except ValueError:
+        return None
+
+
+def get_semver_from_package_manager(package_manager: str) -> semver.version.Version:
+    """Parse packageManager from package.json and return a semver Version if possible.
+
+    :raises UnexpectedFormat:
+        if packageManager doesn't match the name@semver format
+        if packageManager does not specify yarn
+        if packageManager version is not a valid semver
+    """
+    # https://github.com/nodejs/corepack/blob/787e24df609513702eafcd8c6a5f03544d7d45cc/sources/specUtils.ts#L10
+    package_manager_spec_pattern = re.compile(r"^(?!_)(.+)@(.+)$")
+    match = package_manager_spec_pattern.match(package_manager)
+    if not match:
+        raise UnexpectedFormat(
+            "could not parse packageManager spec in package.json (expected name@semver)"
+        )
+
+    name, version = match.groups()
+    if name != "yarn":
+        raise UnexpectedFormat("packageManager in package.json must be yarn")
+
+    try:
+        return semver.version.Version.parse(version)
+    except ValueError as e:
+        raise UnexpectedFormat(
+            f"{version} is not a valid semver for packageManager in package.json"
+        ) from e
```

Comment:
Is semver.version.Version able to parse every valid npm semver version?

If not, we can leave the validation to corepack

Reply:
I tested the usual suspects, but I can do a more extensive verification. The main reason for this check was to ensure that we can do a comparison with the yarnPath semver (if necessary) later in _configure_yarn_version